CMakeLists.txt

```diff
@@ -523,6 +525,29 @@ if(OPENCV_DISABLE_FILESYSTEM_SUPPORT)
   add_definitions(-DOPENCV_HAVE_FILESYSTEM_SUPPORT=0)
 endif()
 
+if(OPENCV_DISABLE_THREAD_SUPPORT)
+  add_definitions(-DOPENCV_DISABLE_THREAD_SUPPORT)
+  # Force removal of code conditionally compiled with `#if
+  # HAVE_PTHREAD`.
+  set(HAVE_PTHREAD 0)
+  # Disable thread/mutex in openjpeg.
+  set(OPJ_USE_THREAD OFF)
+  # There components are disabled because they require
+  # multi-threaded execution.
+  set(WITH_PROTOBUF OFF)
+  set(WITH_GSTREAMER OFF)
+  set(WITH_OPENCL OFF)
+
+  # These modules require `#include
+  # <[thread|mutex|condition_variable|future]>` and linkage into
+  # `libpthread` to work. Notice that disabling `videoio` disables
+  # the `ts` module.
+  set(BUILD_opencv_videoio OFF)
+  set(BUILD_opencv_objdetect OFF)
+  set(BUILD_opencv_gapi OFF)
+  set(BUILD_opencv_dnn OFF)
```

Comment:
Comment about `GTEST_HAS_PTHREAD` is applicable for all these flags too.
Root CMake script should not know implementation details of other modules or dependencies.

Reply:
Just to make sure I don't miss anything: the only option that is left to be moved is `OPJ_USE_THREAD`, right?

Also, am I correct in assuming that I can do the following in the openjpeg's CMakeList.txt?

```
if (OPENCV_DISABLE_THREAD_SUPPORT)
  set(OPJ_USE_THREAD OFF)
endif()
```

(Put it in other words: is `OPENCV_DISABLE_THREAD_SUPPORT` visible in the cmake configuration of openjpg/?)

Reply:
Yes, this option should work.

(please remove space after `if` and `bracket`)

Reply:
Done

Reply:
Hum - `OPENCV_DISABLE_THREAD_SUPPORT` does not seem to be visible in the openjpeg folder.

I get the following message when trying to build for baremetal after moving ` set(OPJ_USE_THREAD OFF)` in openjpeg:

```
/home/frapet01/projects/opencv/opencv/3rdparty/openjpeg/openjp2/thread.c: In function 'opj_thread_join':
/home/frapet01/projects/opencv/opencv/3rdparty/openjpeg/openjp2/thread.c:423:5: warning: implicit declaration of function 'pthread_join'; did you mean 'opj_thread_join'? [-Wimplicit-function-declaration]
  423 |     pthread_join(thread->thread, &status);
      |     ^~~~~~~~~~~~
      |     opj_thread_join
```

Would setting `set(OPJ_USE_THREAD OFF)` in the toolchain file of the baremetal platform OK? Or is there a more canonical way to do this?